File: web/src/lib/teamComputer.ts
import type { AgentTeam, EmployeeAgent } from "../types.js";

/*
 * A team lives on one computer. Team dispatch runs every member in one shared
 * workspace, so setup picks the computer first and the roster is drawn from
 * the agents on it — the same rule a project's crew follows. The backend
 * (`backend/relay/services/team_computer.py`) refuses a roster that spans
 * computers; these helpers keep the form from offering one.
 */

type TeamMembership = { memberIds: string[]; leadId: string };

/** The computers an agent runs on: its active placements, or the computer it
 *  was created on while nothing has been placed yet. Mirrors the backend.
 *
 *  Active only, like every *setup* rule (a project crew's
 *  `validate_project_roster`, the thread picker's `teamsForThreadNode`): a
 *  draining agent is leaving, so a new roster must not be built on it. This is
 *  deliberately stricter than `agentOnComputer` in `taskAssignment.ts`, which
 *  admits work to a roster that already exists. */
export function agentComputerIds(agent: EmployeeAgent): string[] {
  const placed = agent.placements
    .filter((placement) => placement.desiredState === "active" && placement.computerId)
    .map((placement) => placement.computerId as string);
  if (placed.length > 0) return [...new Set(placed)];
  return agent.computerId ? [agent.computerId] : [];
}

function isOnComputer(agent: EmployeeAgent, computerId: string): boolean {
  return agentComputerIds(agent).includes(computerId);
}

/** The agents a team on this computer may take, in roster order. Current
 *  members stay listed even when the computer does not host them, so a
 *  legacy split roster can be seen and trimmed rather than silently hidden. */
export function agentsForTeamComputer(
  agents: readonly EmployeeAgent[],
  computerId: string,
  memberIds: readonly string[],
): EmployeeAgent[] {
  if (!computerId) return [];
  return agents.filter((agent) => !agent.deletedAt
    && (isOnComputer(agent, computerId) || memberIds.includes(agent.id)));
}

/** Members the computer does not host — each one blocks saving. */
export function membersOffComputer(
  agents: readonly EmployeeAgent[],
  computerId: string,
  memberIds: readonly string[],
): string[] {
  return memberIds.filter((id) => {
    const agent = agents.find((candidate) => candidate.id === id);
    return !agent || !isOnComputer(agent, computerId);
  });
}

/** Switching computers keeps only the members the new one hosts; a lead who
 *  is dropped hands the role to the next remaining member. */
export function pruneMembershipToComputer(
  membership: TeamMembership,
  agents: readonly EmployeeAgent[],
  computerId: string,
): TeamMembership {
  const off = new Set(membersOffComputer(agents, computerId, membership.memberIds));
  const memberIds = membership.memberIds.filter((id) => !off.has(id));
  const leadId = memberIds.includes(membership.leadId) ? membership.leadId : (memberIds[0] ?? "");
  return { memberIds, leadId };
}

/** The computer a team lives on: the one it recorded, else — for a team
 *  saved before teams carried one — the computer hosting its whole roster,
 *  else its lead's, so a legacy split team opens on a sensible machine. */
export function teamComputerId(
  team: Pick<AgentTeam, "memberAgentIds"> & Partial<Pick<AgentTeam, "computerId" | "leadAgentId">>,
  agents: readonly EmployeeAgent[],
): string {
  if (team.computerId) return team.computerId;
  const rosters = team.memberAgentIds.map((id) => {
    const agent = agents.find((candidate) => candidate.id === id);
    return agent ? agentComputerIds(agent) : [];
  });
  const shared = (rosters[0] ?? []).filter((id) => rosters.every((roster) => roster.includes(id)));
  if (shared.length > 0) return [...shared].sort()[0];
  const lead = agents.find((agent) => agent.id === team.leadAgentId);
  return lead ? (agentComputerIds(lead)[0] ?? "") : "";
}

File: backend/relay/services/team_computer.py
"""A team lives on one computer.

Team dispatch is a lead-first pipeline across every member in one shared
workspace, so a roster split across machines can never take a thread. Team
setup therefore picks the computer first and draws the roster from the agents
on it; this module is the backend half of that rule, mirroring what
`validate_project_roster` does for a project's crew.
"""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any

from ..core.computer_identity import computer_id
from ..persistence.team_store import TeamValidationError


def placement_on_computer(
    placement: Mapping[str, Any],
    target_computer_id: str,
    target_node_id: str | None,
) -> bool:
    """Is this placement an active one on the target computer?

    Legacy placements predate stable computer ids and only name the daemon
    node, so they match on the node currently hosting the computer.
    """
    if placement.get("desiredState") != "active":
        return False
    if placement.get("computerId"):
        return placement["computerId"] == target_computer_id
    return target_node_id is not None and placement.get("daemonNodeId") == target_node_id


def _agent_computer_ids(agent: Mapping[str, Any], registry: Any, placement_store: Any) -> set[str]:
    """The computers an agent actually runs on: its active placements, or the
    computer it was created on when nothing has been placed yet."""
    found: set[str] = set()
    for placement in placement_store.list_placements(agent_id=agent["id"]):
        if placement.get("desiredState") != "active":
            continue
        if placement.get("computerId"):
            found.add(placement["computerId"])
            continue
        node = registry.get(placement.get("daemonNodeId") or "")
        if node:
            found.add(computer_id(node))
    if not found and agent.get("computerId"):
        found.add(agent["computerId"])
    return found


def _named_computer(owner_employee_id: str, value: Any, registry: Any) -> str:
    """Validate the computer a team write names.

    Matched by stable computer id and owner together, as agent creation does:
    naming another employee's computer id must not pass.
    """
    if not isinstance(value, str) or not value.strip():
        raise TeamValidationError("team_computer_required")
    target = value.strip()
    nodes = [
        node
        for node in registry.monitor_nodes()
        if computer_id(node) == target
        and not node.get("retiredAt")
        and node.get("status") != "deleted"
    ]
    if not nodes:
        raise TeamValidationError("team_computer_not_found")
    if not any(node.get("employeeId") == owner_employee_id for node in nodes):
        raise TeamValidationError("team_computer_forbidden")
    return target


def resolve_team_computer(
    owner_employee_id: str,
    body: Mapping[str, Any],
    member_ids: list[str],
    *,
    current: Mapping[str, Any] | None,
    registry: Any,
    agent_store: Any,
    placement_store: Any,
) -> str | None:
    """Return the computer the team lives on after this write, or None when the
    write changes neither who is on the roster nor the computer.

    The computer is the one the write names, else the team's recorded
    one, else — for a caller that never names one — the single computer
    hosting the whole roster. Every member must be on it.
    """
    names_computer = "computerId" in body
    # Clients resend the stored roster with every save; only a change in who
    # is on the team (order is not placement) can move it off its computer.
    roster_changed = current is None or set(member_ids) != set(
        current.get("memberAgentIds") or []
    )
    if not names_computer and not roster_changed:
        return None
    target = (
        _named_computer(owner_employee_id, body.get("computerId"), registry)
        if names_computer
        else (current or {}).get("computerId")
    )
    rosters = [
        _agent_computer_ids(agent_store.get_agent(agent_id) or {"id": agent_id}, registry, placement_store)
        for agent_id in member_ids
    ]
    if target:
        if any(target not in computers for computers in rosters):
            raise TeamValidationError("team_member_computer_mismatch")
        return target
    shared = set.intersection(*rosters) if rosters else set()
    if not shared:
        raise TeamValidationError("team_member_computer_mismatch")
    # Several shared computers is legal but ambiguous; sorting keeps the
    # choice stable across retries of the same write.
    return sorted(shared)[0]
